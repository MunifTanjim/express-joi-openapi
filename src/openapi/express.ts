import { Application, Router } from 'express'
import { RegisteredExpressOpenAPIPlugin } from '../plugins/types'
import { HttpMethod } from '../types'
import { OpenAPISpecification } from './index'
import { Layer, Route, StackLayer, StackName } from './types'

const regexPattern = {
  param: /\(\?:\(\[\^\\\/]\+\?\)\)/,
  path: /^\^\\\/(?:(:?[\w\\.-]*(?:\\\/:?[\w\\.-]*)*)|(\(\?:\(\[\^\\\/]\+\?\)\)))\\/,
  rootPath: /^\^\\\/\?\(\?=\\\/\|\$\)$/,
}

const getPathString = (pathRegex: RegExp, keys: Layer['keys']): string => {
  const pathRegexpSource = pathRegex.source

  let regexString = pathRegexpSource

  let regexStringExecArray = regexPattern.path.exec(regexString)

  let paramIndex = 0
  while (regexPattern.param.test(regexString)) {
    const param = `:${keys[paramIndex].name}`

    regexString = regexString.replace(regexPattern.param, param)

    paramIndex++
  }

  if (regexString !== pathRegexpSource) {
    regexStringExecArray = regexPattern.path.exec(regexString)
  }

  const pathString = regexStringExecArray![1].replace(/\\\//g, '/')

  return pathString
}

const processRoute = (
  spec: OpenAPISpecification,
  route: Route,
  basePath = '',
  plugins: RegisteredExpressOpenAPIPlugin[]
): void => {
  const routePaths = Array.isArray(route.path) ? route.path : [route.path]

  routePaths.forEach((routePath) => {
    const path = `${
      basePath ? `${basePath}${routePath === '/' ? '' : routePath}` : routePath
    }`.replace(/:([a-zA-Z0-9_]+)/g, '{$1}')

    const methods = Object.keys(route.methods).filter(
      (method): method is HttpMethod => method !== '_all'
    )

    if (!spec.paths[path]) {
      spec.setPathItem(path, {})
    }

    for (const method of methods) {
      spec.setPathItemOperation(path, method, {
        responses: {
          default: {
            description: '',
          },
        },
      })

      for (const plugin of plugins) {
        const stashedValue = plugin.stash.find(route, method)
        if (stashedValue) {
          plugin.processRoute(plugin.specification, stashedValue, {
            path,
            method,
          })
        }
      }
    }
  })
}

export const processExpressRouters = (
  spec: OpenAPISpecification,
  app: Application | Router,
  basePath = '',
  plugins: RegisteredExpressOpenAPIPlugin[]
): void => {
  const stack: StackLayer[] = app.stack ?? (app as Application)._router?.stack

  if (!stack) {
    spec.setPathItem(basePath, {})

    return
  }

  for (const layer of stack) {
    // terminal route
    if (layer.route) {
      processRoute(spec, layer.route, basePath, plugins)

      continue
    }

    if (
      layer.name !== StackName.ROUTER &&
      layer.name !== StackName.BOUND_DISPATCH &&
      layer.name !== StackName.MOUNTED_APP
    ) {
      continue
    }

    if (regexPattern.path.test(layer.regexp.source)) {
      const parsedPath = getPathString(layer.regexp, layer.keys)

      processExpressRouters(
        spec,
        layer.handle as Router,
        basePath + '/' + parsedPath,
        plugins
      )

      continue
    }

    if (
      !layer.path &&
      layer.regexp &&
      !regexPattern.rootPath.test(layer.regexp.source)
    ) {
      const regexPath = ' RegExp(' + layer.regexp + ') '

      processExpressRouters(
        spec,
        layer.handle as Router,
        basePath + '/' + regexPath,
        plugins
      )

      continue
    }

    processExpressRouters(spec, layer.handle as Router, basePath, plugins)
  }
}
