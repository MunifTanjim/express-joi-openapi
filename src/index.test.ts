import Joi from '@hapi/joi'
import express from 'express'
import { initializeJoiOpenApi } from './index'
import request from 'supertest'

describe('initializeJoiOpenApi', () => {
  test('basic request schema generation', () => {
    const {
      getRequestValidationMiddleware,
      getOpenApiSpecification,
    } = initializeJoiOpenApi({ Joi })

    const app = express()

    app.post(
      '/ping',
      getRequestValidationMiddleware({
        query: {
          count: Joi.number()
            .integer()
            .required()
            .label('pong count')
            .description('number of times ping will pong'),
        },
        body: {
          type: Joi.string()
            .valid('plastic', 'rubber', 'wood')
            .default('rubber'),
        },
      })
    )

    const specification = getOpenApiSpecification(app)

    expect(specification.paths).toMatchSnapshot()
  })

  test('basic response schema generation', async () => {
    const {
      getResponseValidationMiddleware,
      getOpenApiSpecification,
    } = initializeJoiOpenApi({ Joi })

    const app = express()

    app.post(
      '/ping',
      getResponseValidationMiddleware({
        200: {
          body: Joi.object({
            pong: Joi.number().integer().required(),
          }),
        },
      }),
      (_req, res) => {
        res.status(200).json({ pong: 42 })
      }
    )

    const specification = getOpenApiSpecification(app)

    expect(specification.paths).toMatchSnapshot()

    const response = await request(app).post('/ping')

    expect(response.status).toBe(200)
    expect(response.body).toMatchObject({ pong: 42 })
  })
})
