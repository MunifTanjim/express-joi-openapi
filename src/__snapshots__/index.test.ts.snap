// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`initializeJoiOpenApi basic request schema generation 1`] = `
Object {
  "/ping": Object {
    "post": Object {
      "parameters": Array [
        Object {
          "in": "query",
          "name": "count",
          "required": true,
          "schema": Object {
            "description": "number of times ping will pong",
            "title": "pong count",
            "type": "integer",
          },
        },
      ],
      "requestBody": Object {
        "content": Object {
          "application/json": Object {
            "schema": Object {
              "additionalProperties": false,
              "properties": Object {
                "type": Object {
                  "default": "rubber",
                  "enum": Array [
                    "plastic",
                    "rubber",
                    "wood",
                  ],
                  "type": "string",
                },
              },
              "type": "object",
            },
          },
        },
      },
      "responses": Object {
        "default": Object {
          "description": "",
        },
      },
    },
  },
}
`;

exports[`initializeJoiOpenApi basic response schema generation 1`] = `
Object {
  "/ping": Object {
    "post": Object {
      "responses": Object {
        "200": Object {
          "content": Object {
            "application/json": Object {
              "schema": Object {
                "additionalProperties": false,
                "properties": Object {
                  "pong": Object {
                    "type": "integer",
                  },
                },
                "required": Array [
                  "pong",
                ],
                "type": "object",
              },
            },
          },
          "description": "",
        },
        "default": Object {
          "description": "",
        },
      },
    },
  },
}
`;
